File: UsingFMDB-Objective-C/UsingFMDB-Objective-C/AppDAO.m
//
//  AppDAO.m
//  UsingFMDB-Objective-C
//
//  Created by akabeko on 2016/12/12.
//  Copyright © 2016年 akabeko. All rights reserved.
//

#import "AppDAO.h"
#import <FMDatabase.h>

/** File name of the SQLite database. */
static NSString * const kDBFileName = @"app.db";

@interface AppDAO ()

/** Path of the SQLite database file. */
@property (nonatomic) NSString *dbFilePath;

@end

@implementation AppDAO

/**
 * Initialize the instance.
 *
 * @return Instance.
 */
- (instancetype)init {
    self = [super init];
    if (self) {
        self.dbFilePath = [self databaseFilePath];
        NSLog(@"DB = %@", self.dbFilePath);
    }

    return self;
}

/**
 * Get the database connection.
 *
 * @return Connection.
 */
- (FMDatabase *)connection {
    FMDatabase* db = [FMDatabase databaseWithPath:self.dbFilePath];
    return ([db open] ? db : nil);
}

#pragma mark - Private

/**
 * Get the path of database file.
 *
 * @return file path.
 */
- (NSString *)databaseFilePath {
    NSArray*  paths = NSSearchPathForDirectoriesInDomains( NSDocumentDirectory, NSUserDomainMask, YES );
    NSString* dir   = [paths objectAtIndex:0];

    return [dir stringByAppendingPathComponent:kDBFileName];
}

@end

File: UsingFMDB-Swift/UsingFMDB-Swift/AppDAO.swift
//
//  AppDAO.swift
//  UsingFMDB-Swift
//
//  Created by akabeko on 2016/12/17.
//  Copyright © 2016年 akabeko. All rights reserved.
//

import UIKit
import FMDB

/// Manager fot the application data.
class AppDAO: NSObject {
    /// Manage for the books data.
    private(set) var bookDAO: BookDAO?

    /// Path of the database file.
    private var filePath = AppDAO.databaseFilePath()

    /// Initialize the instance.
    override init() {
        super.init()
        self.bookDAO = BookDAO(appDAO: self)
        print(self.filePath)
    }

    /// Get the database connection.
    ///
    /// - Returns: Connection.
    func connection() -> FMDatabase? {
        let db = FMDatabase(path: self.filePath)
        return (db?.open())! ? db : nil
    }

    /// Get the path of database file.
    ///
    /// - Returns: Path of the database file.
    private static func databaseFilePath() -> String {
        let paths = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)
        let dir   = paths[0] as NSString
        return dir.appendingPathComponent("app.db")
    }
}
